pymeasure/instruments/channel.py

```diff
@@ -0,0 +1,103 @@
+#
+# This file is part of the PyMeasure package.
+#
+# Copyright (c) 2013-2022 PyMeasure Developers
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+#
+
+import logging
+
+from .commonBase import CommonBase
+
+log = logging.getLogger(__name__)
+log.addHandler(logging.NullHandler())
+
+
+class Channel(CommonBase):
+    """The base class for channel definitions.
+
+    This class supports dynamic properties like :class:`Instrument`,
+    but requires an :class:`Instrument` instance as a parent for communication.
+    The default implementation of :meth:`write` uses :code:`str.format` to
+    exchange :code:`'{ch}'` for the channel name.
+
+    :param parent: The instrument (an instance of :class:`~pymeasure.instruments.Instrument`)
+        to which the channel belongs.
+    :param id: Identifier of the channel, as it is used for the communication.
+    """
+
+    def __init__(self, parent, id):
+        super().__init__()
+        self.parent = parent
+        self.id = id
+
```

Comment:
```suggestion
        # Copy method from parent class
        for method in ('read', 'write_bytes', 'read_bytes', write_binary_values', 'read_binary_values', 'wait_for'):
            setattr(self, method, getattr(self.parent, method))
```
I see also `binary_values` missing. Is it intentional ?

Reply:
I like your suggestion. Thanks!

`binary_values` (similar as `values`) is a method of `CommonBase`, because it does not communicate itself, but calls `write` and `read_binary_values` (similar to `values`).

Reply:
You could check, whether user has overridden the method before overwriting it, but if this make code too complex or not readable, just leave as you have done.

Reply:
> You could check, whether user has overridden the method before overwriting it, but if this make code too complex or not readable, just leave as you have done.

The problem is not the code, it would be just one simple if clause.

More diffcult is, that the implementor would have to write
```
def write(self, command):
    self.parent.write("something"+command)
```
in order to subclass the `write` method in a channel, because the super class does not have a `write` method. How does he know how to use it?

For the instrument, however, he has to write
```
def write(self, command):
    super().write("prefix" + command)
```
which is common practice for subclassing methods.


Another solution for code reduction would be, to base the instrument on the channel, because only the init and the write method would need a change.
However, the instrument would have to invoke the adapter as `parent`, which is really strange. If we want to keep it more descriptive (instrument calling adapter and not parent), we need these extra lines.

Reply:
You are absolutely right. Please leave as it is.